File: common/src/main/java/cc/turtl/chiselmon/feature/pc/PCButton.java
package cc.turtl.chiselmon.feature.pc;

import cc.turtl.chiselmon.util.format.ColorUtils;
import cc.turtl.chiselmon.util.render.TextRenderUtils;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import com.cobblemon.mod.common.CobblemonSounds;
import com.cobblemon.mod.common.client.gui.CobblemonRenderable;

import net.minecraft.client.gui.GuiGraphics;
import net.minecraft.client.gui.components.Button;
import net.minecraft.client.gui.components.Tooltip;
import net.minecraft.client.resources.sounds.SimpleSoundInstance;
import net.minecraft.client.resources.sounds.SoundInstance;
import net.minecraft.client.sounds.SoundManager;
import net.minecraft.network.chat.Component;
import net.minecraft.resources.ResourceLocation;

/**
 * Generic PC button with configurable sprite, tooltip, and optional text rendering.
 * All PC buttons should use this class directly rather than subclassing.
 */
public class PCButton extends Button implements CobblemonRenderable {

    private final ResourceLocation sprite;
    private final int textureWidth;
    private final int textureHeight;
    private final int buttonHeight;

    // Optional text rendering
    private final boolean renderText;
    private final int textColor;
    private final int textMargin;

    // Tooltip configuration
    private final Tooltip activeTooltip;
    private final Tooltip inactiveTooltip;

    // State
    private boolean active;

    private PCButton(Builder builder) {
        super(
                builder.x,
                builder.y,
                builder.width,
                builder.height,
                builder.text != null ? builder.text : Component.empty(),
                builder.onPress,
                DEFAULT_NARRATION
        );

        this.sprite = builder.sprite;
        this.textureWidth = builder.textureWidth;
        this.textureHeight = builder.textureHeight;
        this.buttonHeight = builder.height;

        this.renderText = builder.text != null;
        this.textColor = builder.textColor;
        this.textMargin = builder.textMargin;

        this.activeTooltip = builder.activeTooltip;
        this.inactiveTooltip = builder.inactiveTooltip;

        this.active = builder.initialActive;
    }

    @Override
    protected void renderWidget(@NotNull GuiGraphics context, int mouseX, int mouseY, float partialTicks) {
        // Update tooltip based on state
        updateTooltip();

        // Calculate texture offset based on active state and hover
        int textureYOffset = calculateTextureYOffset();

        // Render the button sprite
        context.blit(
                sprite,
                getX(),
                getY(),
                0,
                textureYOffset,
                width,
                height,
                textureWidth,
                textureHeight
        );

        // Render text if configured
        if (renderText) {
            renderButtonText(context);
        }
    }

    @Override
    public void playDownSound(@NotNull SoundManager soundManager) {
        soundManager.play((SoundInstance) SimpleSoundInstance.forUI(CobblemonSounds.PC_CLICK, 1.0F));
    }

    private int calculateTextureYOffset() {
        // When active: show bottom half when not hovered, top half when hovered
        // When inactive: show top half when not hovered, bottom half when hovered
        return active
                ? (isHovered() ? 0 : buttonHeight)
                : (isHovered() ? buttonHeight : 0);
    }

    private void updateTooltip() {
        if (activeTooltip != null && inactiveTooltip != null) {
            setTooltip(active ? activeTooltip : inactiveTooltip);
        } else if (inactiveTooltip != null) {
            setTooltip(inactiveTooltip);
        }
    }

    private void renderButtonText(GuiGraphics context) {
        int centerX = getX() + width / 2;
        int centerY = getY() + height / 2;

        TextRenderUtils.renderCenteredText(
                context,
                getMessage(),
                textColor,
                centerX,
                centerY,
                width - textMargin
        );
    }

    public boolean isActive() {
        return active;
    }

    public void setActive(boolean active) {
        this.active = active;
    }

    public void toggleActive() {
        this.active = !this.active;
    }

    // Builder pattern for clean construction
    public static class Builder {
        // Required parameters
        private final int x;
        private final int y;
        private final ResourceLocation sprite;
        private final int textureWidth;
        private final int textureHeight;
        private final OnPress onPress;

        // Derived from texture
        private int width;
        private int height;

        // Optional parameters with defaults
        private Component text = null;
        private int textColor = ColorUtils.WHITE;
        private int textMargin = 5;
        private Tooltip activeTooltip = null;
        private Tooltip inactiveTooltip = null;
        private boolean initialActive = false;

        /**
         * Create a button builder.
         *
         * @param x X position
         * @param y Y position
         * @param sprite Texture location
         * @param textureWidth Full texture width
         * @param textureHeight Full texture height (should be 2x button height for hover states)
         * @param onPress Click handler
         */
        public Builder(int x, int y, ResourceLocation sprite, int textureWidth, int textureHeight, OnPress onPress) {
            this.x = x;
            this.y = y;
            this.sprite = sprite;
            this.textureWidth = textureWidth;
            this.textureHeight = textureHeight;
            this.onPress = onPress;

            // By default, button dimensions match texture (height is half for top/bottom states)
            this.width = textureWidth;
            this.height = textureHeight / 2;
        }

        /**
         * Override the button dimensions (useful if texture has padding).
         */
        public Builder dimensions(int width, int height) {
            this.width = width;
            this.height = height;
            return this;
        }

        /**
         * Add text to be rendered on the button.
         */
        public Builder text(Component text) {
            this.text = text;
            return this;
        }

        /**
         * Configure text rendering appearance.
         */
        public Builder textStyle(int color, int margin) {
            this.textColor = color;
            this.textMargin = margin;
            return this;
        }

        /**
         * Set a single tooltip (shown regardless of active state).
         */
        public Builder tooltip(Tooltip tooltip) {
            this.inactiveTooltip = tooltip;
            return this;
        }

        /**
         * Set different tooltips for active and inactive states.
         * Useful for toggle buttons (e.g., "Add bookmark" vs "Remove bookmark").
         */
        public Builder tooltips(Tooltip activeTooltip, Tooltip inactiveTooltip) {
            this.activeTooltip = activeTooltip;
            this.inactiveTooltip = inactiveTooltip;
            return this;
        }

        /**
         * Set the initial active state (default: false).
         */
        public Builder active(boolean active) {
            this.initialActive = active;
            return this;
        }

        public PCButton build() {
            return new PCButton(this);
        }
    }
}

File: common/src/main/kotlin/cc/turtl/chiselmon/client/feature/pc/PCButton.kt
package cc.turtl.chiselmon.client.feature.pc

import cc.turtl.chiselmon.client.util.renderCenteredText
import cc.turtl.turtlshell.api.core.format.ColorLib
import com.cobblemon.mod.common.CobblemonSounds
import com.cobblemon.mod.common.client.gui.CobblemonRenderable
import net.minecraft.client.gui.GuiGraphics
import net.minecraft.client.gui.components.Button
import net.minecraft.client.gui.components.Tooltip
import net.minecraft.client.resources.sounds.SimpleSoundInstance
import net.minecraft.client.sounds.SoundManager
import net.minecraft.network.chat.Component
import net.minecraft.resources.ResourceLocation

class PCButton private constructor(builder: Builder) : Button(
    builder.x, builder.y, builder.width, builder.height,
    builder.text ?: Component.empty(),
    builder.onPress,
    DEFAULT_NARRATION
), CobblemonRenderable {

    private val sprite: ResourceLocation = builder.sprite
    private val textureWidth: Int = builder.textureWidth
    private val textureHeight: Int = builder.textureHeight
    private val buttonHeight: Int = builder.height
    private val renderText: Boolean = builder.text != null
    private val textColor: Int = builder.textColor
    private val textMargin: Int = builder.textMargin
    private val activeTooltip: Tooltip? = builder.activeTooltip
    private val inactiveTooltip: Tooltip? = builder.inactiveTooltip
    private var toggled: Boolean = builder.initialActive

    override fun renderWidget(context: GuiGraphics, mouseX: Int, mouseY: Int, partialTicks: Float) {
        updateTooltip()

        val textureYOffset = if (toggled) {
            if (isHovered) 0 else buttonHeight
        } else {
            if (isHovered) buttonHeight else 0
        }

        context.blit(sprite, x, y, 0F, textureYOffset.toFloat(), width, height, textureWidth, textureHeight)

        if (renderText) {
            renderCenteredText(context, message, textColor, x + width / 2, y + height / 2, width - textMargin)
        }
    }

    override fun playDownSound(soundManager: SoundManager) {
        soundManager.play(SimpleSoundInstance.forUI(CobblemonSounds.PC_CLICK, 1.0f))
    }

    private fun updateTooltip() {
        if (activeTooltip != null && inactiveTooltip != null) {
            tooltip = if (toggled) activeTooltip else inactiveTooltip
        } else {
            inactiveTooltip?.let { tooltip = it }
        }
    }

    fun setActive(toggled: Boolean) {
        this.toggled = toggled
    }

    fun toggleActive() {
        toggled = !toggled
    }

    class Builder(
        val x: Int,
        val y: Int,
        val sprite: ResourceLocation,
        val textureWidth: Int,
        val textureHeight: Int,
        val onPress: OnPress
    ) {
        var width: Int = textureWidth
        var height: Int = textureHeight / 2
        var text: Component? = null
        var textColor: Int = ColorLib.WHITE.rgb
        var textMargin: Int = 5
        var activeTooltip: Tooltip? = null
        var inactiveTooltip: Tooltip? = null
        var initialActive: Boolean = false

        fun dimensions(width: Int, height: Int) = apply { this.width = width; this.height = height }
        fun text(text: Component) = apply { this.text = text }
        fun textStyle(color: Int, margin: Int) = apply { textColor = color; textMargin = margin }
        fun tooltip(tooltip: Tooltip) = apply { inactiveTooltip = tooltip }
        fun tooltips(active: Tooltip, inactive: Tooltip) = apply { activeTooltip = active; inactiveTooltip = inactive }
        fun active(active: Boolean) = apply { initialActive = active }
        fun build() = PCButton(this)
    }
}